metric,tenantcostserver: fix bug in UpdateIfHigher and use counters

tbg opened the pull request:
This PR principally aims to convert a number of "faux gauges" to proper counters. For technical reasons, we have a number of places in the tenantcostserver package which propagate counts from one place into another. The receiving place thus got `.Update()`d, an operation counters did not support. To work around this, these counters were kept as gauges instead, which then confuses DataDog (which doesn't let you take `as_rate()` of a gauge, for example).
This PR introduces an `Update` method and switches everything over to a counter.

In the process of doing so, I also realized that `aggmetric.CounterFloat64`'s `UpdateIfHigher` method wasn't updating the parent at all, meaning the summed up gauges would simply be incorrect. So that bug is now fixed, too.

See https://github.com/cockroachlabs/support/issues/3297 and in particular this bit in the DD notebook:

<img width="1032" alt="image" src="https://github.com/user-attachments/assets/93de5ba4-5939-415e-87ba-b031d9dbe571" />

See also the accompanying [slack thread](https://cockroachlabs.slack.com/archives/C074TGF3QVC/p1746542510980379?thread_ts=1746531269.284379&cid=C074TGF3QVC).

Epic: none

tbg committed:
aggmetric: add (*AggCounter).Update

tbg committed:
aggmetric: fix bug in CounterFloat64.UpdateIfHigher

It wasn't updating `g.parent.g`, meaning that the total would be incorrect.
This affects the following metrics:

- tenant.consumption.request_units
- tenant.consumption.kv_request_units
- tenant.consumption.cross_region_network_ru

Skipping the release note, since tenant-related units are not relevant for
production builds outside of CRL-hosted serverless.

tbg committed:
tenantcostserver: let there be counters

We were previously using Gauges which we updated with values taken from
counters, to work around the fact that counters don't have an "Update"
method. Now they have one, and we can use the proper type.

Notably, the metric will now properly identify itself as a counter,
which improves UX on DataDog.

Touches https://github.com/cockroachlabs/support/issues/3297.

Epic: none

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/146359)
<!-- Reviewable:end -->

RaduBerinde reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

Nice!

We should consider backporting the `UpdateIfHigher` fix if it might affect any other metrics

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/146359#-:-OPkXWISB_Y6GsNWd7jK:bmhlla9)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @aa-joshi, @arjunmahishi, and @dhartunian)

___
*[`pkg/util/metric/aggmetric/counter.go` line 254 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/146359#-OPkXM8wD7Cc0RT4YWa_:-OPkXM8wD7Cc0RT4YWaa:bb79ca8) ([raw file](https://github.com/cockroachdb/cockroach/blob/6383eecded7b2a835b0826c0b7708e48e1f59aaa/pkg/util/metric/aggmetric/counter.go#L254)):*
> ```Go
> 		return
> 	}
> 	g.parent.g.Inc(delta)
> ```

Hm, I wonder if this affects any storage stats too? I know there were some problems with aggregation of store stats but I don't remember the details CC @jbowens 


<!-- Sent from Reviewable.io -->

jbowens reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/146359)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @aa-joshi, @arjunmahishi, @dhartunian, and @RaduBerinde)

___
*[`pkg/util/metric/aggmetric/counter.go` line 254 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/146359#-OPkXM8wD7Cc0RT4YWa_:-OPlWg2PCMurAZs2JWHS:b77sc1o) ([raw file](https://github.com/cockroachdb/cockroach/blob/6383eecded7b2a835b0826c0b7708e48e1f59aaa/pkg/util/metric/aggmetric/counter.go#L254)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

Hm, I wonder if this affects any storage stats too? I know there were some problems with aggregation of store stats but I don't remember the details CC @jbowens
</blockquote></details>

i don't think we use the `aggmetric` pkg for any of our metrics, just the `util/metric.Counter`
___
*[`pkg/util/metric/aggmetric/counter.go` line 135 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/146359#-OPlW8NcC23DTI_uZBrw:-OPlW8NdDxrTb-alc6tB:b3xp8q0) ([raw file](https://github.com/cockroachdb/cockroach/blob/433e60c7c2230db2312fe70250b8fdde30bea294/pkg/util/metric/aggmetric/counter.go#L135)):*
> ```Go
> func (g *Counter) Update(i int64) {
> 	delta := i - atomic.LoadInt64(&g.value)
> 	g.Inc(delta)
> ```

does this type assume external synchronization? otherwise, couldn't two competing goroutines both see the delta and add the same delta?


<!-- Sent from Reviewable.io -->

RaduBerinde reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/146359)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @aa-joshi, @arjunmahishi, @dhartunian, and @jbowens)

___
*[`pkg/util/metric/aggmetric/counter.go` line 135 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/146359#-OPlW8NcC23DTI_uZBrw:-OPlYiyhBB2nKZnzmOjk:bck0oo2) ([raw file](https://github.com/cockroachdb/cockroach/blob/433e60c7c2230db2312fe70250b8fdde30bea294/pkg/util/metric/aggmetric/counter.go#L135)):*
<details><summary><i>Previously, jbowens (Jackson Owens) wrote…</i></summary><blockquote>

does this type assume external synchronization? otherwise, couldn't two competing goroutines both see the delta and add the same delta?
</blockquote></details>

Good point, we should do a CAS


<!-- Sent from Reviewable.io -->

tbg committed:
aggmetric: make concurrent calls to Update atomic

Previously, concurrent callers to update may both
read the old value of the counter in Update, and
cumulatively over-adjust the counters.

RaduBerinde reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/146359)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @aa-joshi, @arjunmahishi, @dhartunian, and @jbowens)

___
*[`pkg/util/metric/aggmetric/counter.go` line 138 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/146359#-OQ90G8X03LpwU4v5abE:-OQ90G8X03LpwU4v5abF:bjq8yn1) ([raw file](https://github.com/cockroachdb/cockroach/blob/bc68db83f8a0c0190d287df4ead3fd26e411785a/pkg/util/metric/aggmetric/counter.go#L138)):*
> ```Go
> // to be frequently Update'd or Inc'd.
> func (g *Counter) Update(i int64) {
> 	var delta int64
> ```

We can just do
```
oldValue := g.value.Swap(newValue)
g.parent.g.Inc(newValue-oldValue)
```


<!-- Sent from Reviewable.io -->

RaduBerinde reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/146359)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @aa-joshi, @arjunmahishi, @dhartunian, @jbowens, and @tbg)

___
*[`pkg/util/metric/aggmetric/counter.go` line 146 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/146359#-OQ90vLp9UNlffoMOcVK:-OQ90vLp9UNlffoMOcVL:b-8ef36b) ([raw file](https://github.com/cockroachdb/cockroach/blob/bc68db83f8a0c0190d287df4ead3fd26e411785a/pkg/util/metric/aggmetric/counter.go#L146)):*
> ```Go
> 		// Raced with concurrent update, try again.
> 	}
> 	g.parent.g.Inc(delta)
> ```

Hm, I don't think we are allowed to call `Inc` with a negative value:

```
func (c *CounterFloat64) Inc(i float64) {
	if buildutil.CrdbTestBuild && i < 0 {
		panic("Counters should not be decremented")
	}
	c.count.Add(i)
}
```
So maybe we'll have to stick to using just `UpdateIfHigher`?


<!-- Sent from Reviewable.io -->

RaduBerinde reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/146359)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @aa-joshi, @arjunmahishi, @dhartunian, @jbowens, and @tbg)

___
*[`pkg/util/metric/aggmetric/counter.go` line 244 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/146359#-OQ91epo7AEWozIORc15:-OQ91epo7AEWozIORc16:b2w9rtf) ([raw file](https://github.com/cockroachdb/cockroach/blob/5d1135dd880d55eeb01e3f6f69217bd4959b7e84/pkg/util/metric/aggmetric/counter.go#L244)):*
> ```Go
> // than the currently set one. It's assumed the caller holds
> func (g *CounterFloat64) UpdateIfHigher(i float64) {
> 	g.value.UpdateIfHigher(i)
> ```

[nit] We could change `g.value.UpdateIfHigher()` and `StoreIfHigher()` to return the previous value (and maybe an `ok bool`)


<!-- Sent from Reviewable.io -->

tbg committed:
aggmetric: change Update to UpdateIfHigher

This avoids incrementing the underlying Counter by a negative
value, which we assert against.

tbg committed:
aggmetric: remove unused method

tbg committed:
aggmetric: improve CounterFloat64.UpdateIfHigher

tbg reviewed (commented):
TFTR!
Can you give it another look since I took your suggestion(s)?

> Dismissed @jbowens from a discussion.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/146359#-:-OQ95YRqEi2MvyOZGUGn:b-aqdxor)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @aa-joshi, @arjunmahishi, and @dhartunian)

___
*[`pkg/util/metric/aggmetric/counter.go` line 146 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/146359#-OQ90vLp9UNlffoMOcVK:-OQ92bDC9ETns9iNn7E9:b-wxd9ay) ([raw file](https://github.com/cockroachdb/cockroach/blob/bc68db83f8a0c0190d287df4ead3fd26e411785a/pkg/util/metric/aggmetric/counter.go#L146)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

Hm, I don't think we are allowed to call `Inc` with a negative value:

```
func (c *CounterFloat64) Inc(i float64) {
	if buildutil.CrdbTestBuild && i < 0 {
		panic("Counters should not be decremented")
	}
	c.count.Add(i)
}
```
So maybe we'll have to stick to using just `UpdateIfHigher`?
</blockquote></details>

Good point. Changed to `UpdateIfHigher` - unfortunately that means I can't adopt your other suggestion to use `Swap`.
___
*[`pkg/util/metric/aggmetric/counter.go` line 244 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/146359#-OQ91epo7AEWozIORc15:-OQ9590M3Lp3H0IYNzRP:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/5d1135dd880d55eeb01e3f6f69217bd4959b7e84/pkg/util/metric/aggmetric/counter.go#L244)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

[nit] We could change `g.value.UpdateIfHigher()` and `StoreIfHigher()` to return the previous value (and maybe an `ok bool`)
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

RaduBerinde reviewed (commented):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/146359#-:-OQ9KbqJBPdt2_WqebMg:bnfp4nl)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @aa-joshi, @arjunmahishi, and @dhartunian)

___
*[`pkg/util/syncutil/atomic.go` line 51 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/146359#-OQ9KSon8c6AJlABc1GF:-OQ9KSon8c6AJlABc1GG:b-7qa5kr) ([raw file](https://github.com/cockroachdb/cockroach/blob/293c175873cb9fd647c4da132981e33a70438946/pkg/util/syncutil/atomic.go#L51)):*
> ```Go
> 		oldInt := f.val.Load()
> 		oldFloat := math.Float64frombits(oldInt)
> 		if oldFloat > new {
> ```

[nit] >=
___
*[`pkg/util/metric/aggmetric/counter.go` line 129 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/146359#-OQ9KAb32P-Ag5GyxrW0:-OQ9KAb32P-Ag5GyxrW1:b-i95s7d) ([raw file](https://github.com/cockroachdb/cockroach/blob/293c175873cb9fd647c4da132981e33a70438946/pkg/util/metric/aggmetric/counter.go#L129)):*
> ```Go
> func (g *Counter) Inc(i int64) {
> 	g.parent.g.Inc(i)
> 	atomic.AddInt64(&g.value, i)
> ```

[nit] Since you're here, can you change `value` to `atomic.Int64`?
___
*[`pkg/util/metric/aggmetric/counter.go` line 137 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/146359#-OQ9KLSY1zVzjB463P1M:-OQ9KLSY1zVzjB463P1N:bb7whx5) ([raw file](https://github.com/cockroachdb/cockroach/blob/293c175873cb9fd647c4da132981e33a70438946/pkg/util/metric/aggmetric/counter.go#L137)):*
> ```Go
> // so it should only be used if the Counter is not expected
> // to be frequently Update'd or Inc'd.
> func (g *Counter) UpdateIfHigher(i int64) {
> ```

[nit] s/i/newValue
___
*[`pkg/util/metric/aggmetric/counter.go` line 263 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/146359#-OQ9KNa53aCGa5wVaSsF:-OQ9KNa53aCGa5wVaSsG:bb7whx5) ([raw file](https://github.com/cockroachdb/cockroach/blob/293c175873cb9fd647c4da132981e33a70438946/pkg/util/metric/aggmetric/counter.go#L263)):*
> ```Go
> // UpdateIfHigher sets the counter's value only if it's higher
> // than the currently set one.
> func (g *CounterFloat64) UpdateIfHigher(i float64) {
> ```

[nit] s/i/newValue


<!-- Sent from Reviewable.io -->

tbg committed:
aggmetric: more optics improvements

tbg committed:
aggmetric: convert int64 to atomic.Int64

tbg commented:
TFTR!

bors r+
